### PR TITLE
8257500: Drawing MultiResolutionImage with ImageObserver "leaks" memory

### PR DESCRIPTION
**Pull request by mrserb**
This bug was reported as a leak of the ImageObserver when the user draws a multiresolution image.

The multiresolution image is an image that contains a few images inside, when the app uses the observer to get notifications about image loading we create a special internal observer and use it to track loading resolution-variants. This internal observer forwards notification from the resolution variant to the application observer.

The mapping from the application observer to the internal observer is done via "SoftCache" which is a kind of HashMap that uses soft references to the key and value.

Here the bug comes, the soft references are cleared only under memory pressure, and unused objects may sit hours in memory before being cleaned. Moreover, the internal observer was implemented using a strong reference to the application observer(it is not obvious since the lambda is used). So the key object refers to the application's observer cannot be clear fast. This causes an even longer delay of the memory cleanup, which was considered by the use as a "leak".

The fix changes the usage of SoftCache to the WeakHashMap, so the key(the application observer) will be cleared when the application lost the reference to it.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257500](https://bugs.openjdk.java.net/browse/JDK-8257500): Drawing MultiResolutionImage with ImageObserver "leaks" memory

### Reviewers
 * [Alexander Zvegintsev](https://openjdk.java.net/census#azvegint) (@azvegint - **Reviewer**)
 * [Alexey Ivanov](https://openjdk.java.net/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2711/head:pull/2711`
`$ git checkout pull/2711`

### CODE REVIEW

File: src/java.desktop/share/classes/sun/awt/image/MultiResolutionToolkitImage.java [hunk 1]
```diff
@@ -103,38 +100,53 @@ public static ImageObserver getResolutionVariantObserver(
...
+                                   int x, int y, int width, int height) {
+            ImageObserver observer = observerRef.get();
+            Image image = imageRef.get();
 
-                            if(concatenateInfo){
-                                flags &= ((ToolkitImage) image).
-                                        getImageRep().check(null);
-                            }
+            if (observer == null || image == null) {
+                return false;
```
Comment (azvegint) [inline src/java.desktop/share/classes/sun/awt/image/MultiResolutionToolkitImage.java:130]: How about a case when `ObserverCache`'s class instance is not collected by GC, but its `observerRef` or `imageRef`  is collected. Is it possible?
If so this class instance will permanently return false on `imageUpdate()` call.

Comment (mrserb) [reply src/java.desktop/share/classes/sun/awt/image/MultiResolutionToolkitImage.java:130]: observerRef points to the same object as a key in this WeakHashMap, so if observerRef will point to null means that the key is already collected->at some point(when any of the methods of this collection will be called) the value of this map(the ObserverCache object) will be collected as well.

**Review by azvegint — APPROVED** [review]

**Review by aivanov-jdk — APPROVED** [review]
Can `MultiResolutionToolkitImageTest.java` not be run on other platforms?

**Comment by mrserb** [comment]
> Can `MultiResolutionToolkitImageTest.java` not be run on other platforms?

The MultiResolutionToolkitImage is responsible to support the "@2x.png" files extension to make an image from a few image files, and currently this only supported on macOS.

**Pull request closed without merging**